### PR TITLE
[EC-117] Resume blockchain download on app restart

### PR DESCRIPTION
**Pull request by LukasGasior1**
Add storage for most recent fully downloaded (with body and receipts) block. Use that to resume fast sync where we left off.

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala [hunk 1]
```diff
@@ -32,6 +37,20 @@ class FastSyncBlockBodiesRequestHandler(
     cleanupAndStop()
   }
 
+  private def updateBestBlockIfNeeded(receivedHashes: Seq[ByteString]): Unit = {
+    val fullBlocks = receivedHashes.flatMap { hash =>
+      for {
+        header <- blockchain.getBlockHeaderByHash(hash)
+        _ <- blockchain.getReceiptsByHash(hash)
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: I don't understand this line :confused:

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: we have just received a portion of block bodies. I want to check what is the (possibly new) highest fully downloaded block (i.e the one that we already have a receipt for) - that's what this piece does.

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: ensure that receipt is in DB?

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: @LukasGasior1 makes sense, thx for clarification :smile:

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncBlockBodiesRequestHandler.scala]: @adamsmo yes, ensure it's in db

**Comment by LukasGasior1** [comment]
Conflicts: resolved
Base branch: updated

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala [hunk 2]
```diff
@@ -32,6 +37,22 @@ class FastSyncReceiptsRequestHandler(
...
+  private def updateBestBlockIfNeeded(receivedHashes: Seq[ByteString]): Unit = {
+    val fullBlocks = receivedHashes.flatMap { hash =>
+      for {
+        header <- blockchain.getBlockHeaderByHash(hash)
+        _ <- blockchain.getBlockBodyByHash(hash)
+      } yield header
+    }
+
+    if (fullBlocks.nonEmpty) {
+      val bestReceivedBlock = fullBlocks.maxBy(_.number)
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala]: @LukasGasior1 is it possible this case isn't covered by tests?

File: src/main/scala/io/iohk/ethereum/db/components/StoragesComponent.scala [hunk 3]
```diff
@@ -5,23 +5,24 @@ import io.iohk.ethereum.domain.BlockchainStorages
 
 trait StoragesComponent {
 
-  val storages: Storages
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/db/components/StoragesComponent.scala]: Leave these as vals please.

File: src/main/scala/io/iohk/ethereum/domain/Blockchain.scala [hunk 4]
```diff
@@ -181,13 +181,13 @@ class BlockchainImpl(
...
 
 trait BlockchainStorages {
-  val blockHeadersStorage: BlockHeadersStorage
-  val blockBodiesStorage: BlockBodiesStorage
-  val blockNumberMappingStorage: BlockNumberMappingStorage
-  val receiptStorage: ReceiptStorage
-  val evmCodeStorage: EvmCodeStorage
-  val mptNodeStorage: MptNodeStorage
-  val totalDifficultyStorage: TotalDifficultyStorage
+  def blockHeadersStorage: BlockHeadersStorage
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/domain/Blockchain.scala]: Leave as vals please.

File: src/main/scala/io/iohk/ethereum/network/FastSyncHost.scala [hunk 5]
```diff
@@ -10,16 +10,16 @@ import io.iohk.ethereum.network.p2p.messages.PV63._
 import io.iohk.ethereum.network.rlpx.RLPxConnectionHandler
 
 trait FastSyncHost {
-  val peerConfiguration: PeerConfiguration
-  val storage: Blockchain
+  def peerConfiguration: PeerConfiguration
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/FastSyncHost.scala]: Leave as vals please.

**Comment by LukasGasior1** [comment]
@mcsherrylabs changed defs back to vals.

Also it's worth noting that the implementation in this PR is going to be changed as in some cases it may not work as expected. Jan and I are currently working on it.

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala (hunk 2 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncReceiptsRequestHandler.scala]: it is possible. however, see my comment below. the implementation is going to change

**Pull request merged**